### PR TITLE
feat(ctb): Start bisection game

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Starts the bisection game for the fault proof alpha, reworks the old storage format to be a flatly stored DAG and fully implements bisection and the original single-step resolution function.

This PR seeks to find a solution for the storage layout spike.

**Metadata**
Closes CLI-4077
Closes CLI-4100
Closes CLI-4105

### CODE REVIEW

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5984** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/monorepo/5984" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/monorepo/5984?utm_source=stack-comment).

File: packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol [hunk 1]
```diff
@@ -0,0 +1,394 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.15;
+
+import { IDisputeGame } from "./interfaces/IDisputeGame.sol";
+import { IVersioned } from "./interfaces/IVersioned.sol";
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol:5]: This looks unused, contract impl should inherit

File: packages/contracts-bedrock/contracts/dispute/lib/LibPosition.sol [hunk 2]
```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.15;
+
+import "../../libraries/DisputeTypes.sol";
+
+/**
+ * @title LibPosition
+ * @author clabby <https://github.com/clabby>
+ * @notice This library contains helper functions for working with the `Position` type.
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/dispute/lib/LibPosition.sol:9]: How can i best verify the correctness of this? Is the spec for this documented someplace?

Comment (clabby) [reply packages/contracts-bedrock/contracts/dispute/lib/LibPosition.sol:9]: No spec just yet, but can work on it. For the important pieces (attack / defend), here are some visuals:

attack
defend

File: packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol (hunk 1 above)
Comment (clabby) [reply packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol:5]: `IVersioned` must be in scope for the `@inheritdoc` natspec tag: https://github.com/ethereum-optimism/optimism/pull/5984/files/a5770fe20e6ecf295d964a62741142c7e1e68df3#diff-3fdd2c2fb84eee383095af3282dfbba9a5563bc58957cdedda64ac85808e788fR389

File: packages/contracts-bedrock/contracts/libraries/DisputeErrors.sol [hunk 3]
```diff
@@ -39,6 +39,27 @@ error CannotDefendRootClaim();
...
+ */
+error ParentDoesNotExist();
+
+/**
+ * @notice Thrown when a move is attempted to be made after the clock has timed out.
+ */
+error ClockTimeExceeded();
+
+/**
+ * @notice Thrown when a move is attempted to be made at the max depth of the game.
```
Comment (refcell) [inline packages/contracts-bedrock/contracts/libraries/DisputeErrors.sol:64]: Nit
```suggestion
 * @notice Thrown when a move is attempted to be made at or greater than the max depth of the game.
```

File: packages/contracts-bedrock/contracts/test/FaultDisputeGame.t.sol [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+ *⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⡤⢯⡌⡿⡇⠘⡷⠀⠁⠀⠀⢀⣰⠢⠲⠛⣈⣸⠦⠤⠶⠴⢬⣐⣊⡂⠀
+ *⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⣤⡪⡗⢫⠞⠀⠆⣀⠻⠤⠴⠐⠚⣉⢀⠦⠂⠋⠁⠀⠁⠀⠀⠀⠀⢋⠉⠇⠀
+ *⠀⠀⠀⠀⣀⡤⠐⠒⠘⡹⠉⢸⠇⠸⠀⠀⠀⠀⣀⣤⠴⠚⠉⠈⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠼⠀⣾⠀
+ *⠀⠀⠀⡰⠀⠉⠉⠀⠁⠀⠀⠈⢇⠈⠒⠒⠘⠈⢀⢡⡂⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢰⠀⢸⡄
+ *⠀⠀⠸⣿⣆⠤⢀⡀⠀⠀⠀⠀⢘⡌⠀⠀⣀⣀⣀⡈⣤⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⠀⢸⡇
+ *⠀⠀⢸⣀⠀⠉⠒⠐⠛⠋⠭⠭⠍⠉⠛⠒⠒⠒⠀⠒⠚⠛⠛⠛⠩⠭⠭⠭⠭⠤⠤⠤⠤⠤⠭⠭⠉⠓⡆
+ *⠀⠀⠘⠿⣷⣶⣤⣤⣀⣀⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣠⣤⣄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⡇
+ *⠀⠀⠀⠀⠀⠉⠙⠛⠛⠻⠿⢿⣿⣿⣷⣶⣶⣶⣤⣤⣀⣁⣛⣃⣒⠿⠿⠿⠤⠠⠄⠤⠤⢤⣛⣓⣂⣻⡇
+ *⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠉⠉⠉⠙⠛⠻⠿⠿⠿⢿⣿⣿⣿⣷⣶⣶⣾⣿⣿⣿⣿⠿⠟⠁
+ *⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠈⠉⠉⠉⠉⠁⠀⠀⠀⠀⠀⠀
```
Comment (refcell) [inline packages/contracts-bedrock/contracts/test/FaultDisputeGame.t.sol:199]: (:
```suggestion
 *⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠈⠉⠉⠉⠉⠁⠀⠀⠀⠀⠀
```

File: packages/contracts-bedrock/contracts/test/LibClock.t.sol [hunk 5]
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.15;
+
+import { Test } from "forge-std/Test.sol";
+import { LibClock } from "../dispute/lib/LibClock.sol";
+import "../libraries/DisputeTypes.sol";
+
+/**
+ * @notice Tests for `LibClock`
+ */
```
Comment (refcell) [inline packages/contracts-bedrock/contracts/test/LibClock.t.sol:10]: I think test contract docstrings aren't super useful and just add LOC. I'd prefer if we removed at least this top level comment

File: packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+    /**
+     * @inheritdoc IFaultDisputeGame
+     */
+    function step(
+        uint256 prestateIndex,
+        uint256 parentIndex,
+        bytes calldata stateData,
+        bytes calldata
+    ) external {
+        // TODO - Call the VM to perform the execution step.
```
[the remark this answers is not in the document]
Comment (clabby) [reply packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol:110]: No interface yet; Cannon contracts are not final. The interface at the moment is:

`function step(bytes32 stateHash, bytes calldata stateData, bytes calldata proof) external returns (bytes32);`

File: packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+     * @notice The current Semver of the FaultDisputeGame implementation.
+     */
+    string internal constant VERSION = "0.0.1";
+
+    /**
+     * @notice The max depth of the game.
+     * @dev TODO: Update this to the value that we will use in prod. Do we want to have the factory
+     *            set this value? Should it be a constant?
+     */
+    uint256 internal constant MAX_GAME_DEPTH = 4;
```
[the remark this answers is not in the document]
Comment (clabby) [reply packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol:37]: I don't know about this - would require the game to know about the `SystemConfig` + create the need for an additional external call. The game depth should not be configurable. Down to throw it in the constants library, though.

File: packages/contracts-bedrock/contracts/dispute/interfaces/IFaultDisputeGame.sol [hunk 8]
```diff
@@ -10,94 +10,66 @@ import { IDisputeGame } from "./IDisputeGame.sol";
...
      * @notice Perform the final step via an on-chain fault proof processor
      * @dev This function should point to a fault proof processor in order to execute
      *      a step in the fault proof program on-chain. The interface of the fault proof
      *      processor contract should be generic enough such that we can use different
      *      fault proof VMs (MIPS, RiscV5, etc.)
-     * @param disagreement The ClaimHash of the disagreement
+     * @param prestateIndex The index of the prestate of the step within `claimData`.
+     * @param parentIndex The index of the parent claim within `claimData`.
+     * @param stateData The stateData of the step is the preimage of the claim @ `prestateIndex`
+     * @param proof ...
```
[the remark this answers is not in the document]
Comment (clabby) [reply packages/contracts-bedrock/contracts/dispute/interfaces/IFaultDisputeGame.sol:67]: cc @protolambda, what is this proof for?

File: packages/contracts-bedrock/contracts/dispute/interfaces/IFaultDisputeGame.sol [hunk 9]
```diff
@@ -10,94 +10,66 @@ import { IDisputeGame } from "./IDisputeGame.sol";
  * @notice The interface for a fault proof backed dispute game.
  */
 interface IFaultDisputeGame is IDisputeGame {
+    /**
+     * @notice The `ClaimData` struct represents the data associated with a Claim.
+     * @dev TODO: Pack `Clock` and `Position` into the same slot. Should require 4 64 bit arms.
+     * @dev TODO: Add bond ID information.
```
[the remark this answers is not in the document]
Comment (clabby) [reply packages/contracts-bedrock/contracts/dispute/interfaces/IFaultDisputeGame.sol:16]: Won't add bond ID information yet, but we can pack `Clock` and `Position`.

File: packages/contracts-bedrock/contracts/test/LibClock.t.sol (hunk 5 above)
Comment (refcell) [reply packages/contracts-bedrock/contracts/test/LibClock.t.sol:10]: ye it's not a big deal either way, we can leave

File: packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (refcell) [reply packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol:37]: Would it though? Couldn't we pass it through as an immutable

[the remark this answers is not in the document]
Comment (tynes) [reply packages/contracts-bedrock/contracts/dispute/FaultDisputeGame.sol:37]: I don't think that we would need to configure this value dynamically so I don't think we need to put it in the sys config

File: packages/contracts-bedrock/contracts/dispute/interfaces/IFaultDisputeGame.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (protolambda) [reply packages/contracts-bedrock/contracts/dispute/interfaces/IFaultDisputeGame.sol:57]: proof is for memory access (read/write). State is fixed-size, includes a commitment to memory tree. proof is dynamic-size, includes one or two binary merkle proofs to memory leaf nodes. At least in Cannon. Other VMs might look different

File: packages/contracts-bedrock/contracts/dispute/lib/LibPosition.sol [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        // Shift the high-order 64 bits into the low-order 64 bits, leaving only the `depth`.
+        assembly {
+            _depth := shr(0x40, position)
+        }
+    }
+
+    /**
+     * @notice Pulls the `indexAtDepth` out of a packed `Position` type.
+     */
+    function indexAtDepth(Position position) internal pure returns (uint64 _indexAtDepth) {
```
Comment (refcell) [inline packages/contracts-bedrock/contracts/dispute/lib/LibPosition.sol:30]: Are we sticking with underscore suffix named returns?
Then this should be
```suggestion
    function indexAtDepth(Position position) internal pure returns (uint64 indexAtDepth_) {
```

**Review by refcell — CHANGES_REQUESTED** [review]

**Review by tynes — APPROVED** [review]

**Pull request merged**